dev/crypto.md

```diff
@@ -91,29 +89,144 @@ for the Byzantine Fault Tolerance protocol, and (2) rerandomize its
+positions 2 and 3; and so on) and hashes each pair to produce a single
+value in the reduced list (respectively, at position 0; at position
+1; and so on).  To hash two values into a single value, the reduction
+procedure concatenates the domain-separation prefix "MA" together with
+the two values (in the order they appear in the list), and then applies
+the hash function.  When a list has an odd number of values, the last
+value is paired together with an all-zero value (i.e., 32 zero bytes).
+
+The pseudocode for the commitment algorithm is as follows:
+
+```
+def commit(elems):
+  hashes = [H(elem) for elem in elems]
+  return reduce(hashes)
+
+def reduce(hashes):
+  if len(hashes) == 0:
+    return [0 for _ in range(32)]
+  if len(hashes) == 1:
+    return hashes[0]
+  nexthashes = []
+  while len(hashes) > 0:
+    left = hashes[0]
+    right = hashes[1] if len(hashes) > 1 else [0 for _ in range(32)]
+    hashes = hashes[2:]
+    nexthashes.append(H("MA" + left + right))
+  return reduce(nexthashes)
+```
+
+## Proofs
+
+Logically, to verify that an element appears at some position P in the
+array, the verifier runs a variant of the commit procedure to compute
+a candidate root hash, and then checks if the resulting root hash is
+equal to the expected commitment value.  The key difference is that the
+verifier does not have access to the entire list of committed elements;
+the verifier just has some subset of elemens (one or more), along with the
+positions at which these elements appear.  Thus, the verifier needs to
+know the siblings (the `left` and `right` values used in the `reduce()`
+function above) to compute its candidate root hash.  The list of these
+siblings constitutes the proof; thus, a proof is a list of zero or more
+32-byte hash values.  Algorand defines a deterministic order in which
+the verification procedure expects to find siblings in this proof, so no
+additional information is required as part of the proof (in particular,
+no information about which part of the Merkle tree each proof element
+corresponds to).
+
+## Verifying a proof
+
+The following pseudocode defines the logic for verifying a proof (a
+list of 32-byte hashes) for one or more elements, specified as a list
+of position-element pairs, sorted by position in the array, against a
+root commitment.  The function `verify` returns `True` if `proof` is a
+valid proof for all elements in `elems` being present at their positions
+in the array committed to by `root`.  (The pseudocode might raise an
+exception due to accessing the proof past the end; this is equivalent
+to returning `False`.)  The function implements a variant of `reduce()`
+for a sparse array, rather than a fully-populated one.
+
+```
+def verify(elems, proof, root):
+  if len(elems) == 0:
+    return len(proof) == 0
+  if len(elems) == 1 and len(proof) == 0:
+    return elems[0].pos == 0 && elems[0].hash == root
+
+  i = 0
+  nextelems = []
+  while i < len(elems):
```

Comment:
Doesn't this loop infinitely if the condition of the `if` below is not satisfied?
In particular, if `len(elems) = 1`?

Reply:
Good catch; thanks.  I missed incrementing `i` on each loop iteration of this pseudocode.